resources/data-provider.mjs

```diff
@@ -0,0 +1,149 @@
+import { params } from "./shared/params.mjs";
+
+const DEFAULT_TAGS = ["all", "default", "experimental"];
+const DISALLOWED_DOMAINS = ["browserbench.org"];
+export class DataProvider {
+    _tags = new Set(DEFAULT_TAGS);
+    _suites = [];
+
+    get tags() {
+        return this._tags;
+    }
+
+    get suites() {
+        return this._suites;
+    }
+
+    /**
+     * Checks if a given string is a valid URL, supporting both absolute and relative paths.
+     *
+     * This function attempts to construct a URL object. For relative paths, it uses
+     * a dummy base URL to allow the URL constructor to parse them successfully.
+     *
+     * @param {string} url The URL string to validate.
+     * @returns {boolean} True if the URL is valid (absolute or relative), false otherwise.
+     */
+    _isValidUrl(url) {
+        if (typeof url !== "string" || url.length === 0)
+            return false;
+
+        try {
+            new URL(url, "http://www.example.com");
+            return true;
+        } catch (error) {
+            return false;
+        }
+    }
+
+    _freezeSuites() {
+        Object.freeze(this._suites);
+        this._suites.forEach((suite) => {
+            if (!suite.tags)
+                suite.tags = [];
+            if (suite.url.startsWith("experimental/"))
+                suite.tags.unshift("all", "experimental");
+            else
+                suite.tags.unshift("all");
+            suite.enabled = suite.tags.includes("default");
+            Object.freeze(suite.tags);
+            Object.freeze(suite.steps);
+        });
+    }
+
+    _freezeTags() {
+        Object.freeze(this._tags);
+    }
+
+    async init() {
+        if (params.config) {
+            try {
+                const configUrl = new URL(params.config);
+                // Don't fetch if the URL is from DISALLOWED_DOMAINS
+                if (DISALLOWED_DOMAINS.includes(configUrl.hostname)) {
+                    console.warn(`Configuration fetch from a disallowed domain: ${params.config}. Loading default suites.`);
+                    this._loadDefaultSuites();
+                    return;
+                }
+
+                const response = await fetch(params.config);
+                // ✅ Validate that the network request was successful
+                if (!response.ok)
+                    throw new Error(`Could not fetch config: ${response.status}`);
+
+                const config = await response.json();
+                // ✅ Validate the structure of the fetched config object
+                if (!config || !Array.isArray(config.suites))
+                    throw new Error("Could not find a valid config structure!");
+
+                config.suites.flatMap((suite) => suite.tags || []).forEach((tag) => this._tags.add(tag));
+                config.suites.forEach((suite) => {
+                    // ✅ Validate each suite object before processing
+                    if (suite && suite.url && this._isValidUrl(suite.url))
```

Comment:
Why doesn't this result in an exception instead?
Silently skipping a suite when its URL is invalid sounds rather misleading.